maggma/examples/builders.py

```diff
@@ -0,0 +1,274 @@
+        index_warning = (
+            "Missing one or more important indices on stores. "
+            "Performance for large stores may be severely degraded. "
+            "Ensure indices on target.key and "
+            "[(store.lu_field, -1), (store.key, 1)] "
+            "for each of source and target."
+        )
+        logger.warning(index_warning)
+
+    criteria = {}
+    if query:
+        criteria.update(query)
+    if incremental:
+        logger.info(
+            "incremental mode: finding new/updated source keys")
+        keys_updated = source_keys_updated(source, target)
+        # Merge existing criteria and {source.key: {"$in": keys_updated}}.
+        if "$and" in criteria:
+            criteria["$and"].append({source.key: {"$in": keys_updated}})
+        elif source.key in criteria:
+            # XXX could go deeper and check for $in, but this is fine.
+            criteria["$and"] = [
+                {source.key: criteria[source.key].copy()},
+                {source.key: {"$in": keys_updated}},
+            ]
+            del criteria[source.key]
+        else:
+            criteria.update({source.key: {"$in": keys_updated}})
+    # Check ratio of criteria size to 16 MB MongoDB document size limit.
+    # Overestimates ratio via 1000 * 1000 instead of 1024 * 1024.
+    # If criteria is > 16MB, even cursor.count() will fail with a
+    # "DocumentTooLarge: "command document too large" error.
+    if (total_size(criteria) / (16 * 1000 * 1000)) >= 1:
+        raise RuntimeError(
+            "`get_items` query criteria too large. This can happen if "
+            "trying to run incremental=True for the initial build of a "
+            "very large source store, or if `query` is too large. You "
+            "can use maggma.utils.total_size to ensure `query` is smaller "
+            "than 16,000,000 bytes.")
+    return criteria
+
+
+class MapBuilder(Builder, metaclass=ABCMeta):
+    """
+    Apply a unary function to yield a target document for each source document.
+
+    Supports incremental building, where a source document gets built only if it
+    has newer (by lu_field) data than the corresponding (by key) target
+    document.
+
+    """
+    def __init__(self, source, target, query=None, incremental=True, **kwargs):
+        """
+        Apply a unary function to each source document.
+
+        Args:
+            source (Store): source store
+            target (Store): target store
+            query (dict): optional query to filter source store
+            incremental (bool): whether to use lu_field of source and target
+                to get only new/updated documents.
+        """
+        self.source = source
+        self.target = target
+        self.incremental = incremental
+        self.query = query
+        super().__init__(sources=[source], targets=[target], **kwargs)
+        self.kwargs = kwargs.copy()
+        self.kwargs.update(query=query, incremental=incremental)
+
+    @staticmethod
```

Comment:
why static? Can this and the following static functions be made into properties instead?

Reply:
Having it as a method gives a nice error message for free when someone tries to instantiate a subclass without implementing `ufn`. Otherwise, we'd need a `hasattr` check in `__init__` to raise an exception.